prediction_market_agent/agents/microchain_agent/microchain_generators.py

```diff
@@ -0,0 +1,125 @@
+import typing as t
+from enum import Enum
+
+from microchain.models.generators import TokenTracker
+
+
+class Llama31SupportedRole(str, Enum):
+    system = "system"
+    assistant = "assistant"
+    user = "user"
+    ipython = "ipython"
+
+
+class Llama31Message(t.TypedDict):
+    role: Llama31SupportedRole
+    content: str
+
+
+MESSAGE_BLOCK_TEMPLATE = """<|start_header_id|>{role}<|end_header_id|>
+
+{content}
+
+You are a helpful assistant<|eot_id|>"""
+
+
+def verify_system_message_is_first(messages: list[Llama31Message]) -> None:
+    if any(message["role"] == Llama31SupportedRole.system for message in messages[1:]):
+        raise ValueError(
+            f"System message should be the first message in the conversation."
+        )
+
+
+def format_llama31_prompt(messages: list[Llama31Message]) -> str:
```

Comment:
HF `AutoTokenizer` has a util function to do this templating. e.g. for this case:

```
from transformers import AutoTokenizer

messages = [
    {"role": "user", "content": "What's the capital of France'?"},
    {"role": "assistant", "content": "It's Paris."},
]
tokenizer = AutoTokenizer.from_pretrained("meta-llama/Meta-Llama-3.1-405B-Instruct")
encoded_messages = tokenizer.apply_chat_template(messages, tokenize=False)
print(encoded_messages)
```

gives:
```
<|begin_of_text|><|start_header_id|>user<|end_header_id|>

What's the capital of France'?<|eot_id|><|start_header_id|>assistant<|end_header_id|>

It's Paris.<|eot_id|><|start_header_id|>assistant<|end_header_id|>
```

unfortunately for these meta models you need to apply for access on https://huggingface.co/meta-llama/Meta-Llama-3.1-405B-Instruct and sign in via CLI first. So not that useful :(

buuuuut...

if you root around the huggingface 'trending models' page, you can find a finetine/quantization of the model with the correct tokenizer. Not all of them have the correct one (e.g. `"unsloth/Meta-Llama-3.1-8B-bnb-4bit"` has an incorrect one, but `"hugging-quants/Meta-Llama-3.1-70B-Instruct-AWQ-INT4"` matches the official meta one!!

Not sure if you prefer that over what you've done here...

Reply:
Will let you decide if you want to do anything about this bit

Reply:
This is good! It also solves the problem with counting tokens. I'm afraid of calling the transformer's API on each cronjob start, but we can upload these files to our own storage if it becomes a problem.